traces_sampler.go

```diff
@@ -0,0 +1,56 @@
+package sentry
+
+import (
+	"fmt"
+
+	"github.com/getsentry/sentry-go/internal/crypto/randutil"
+)
+
+// A TracesSampler makes sampling decisions for spans.
+//
+// In addition to the sampling context passed to the Sample method,
+// implementations may keep and use internal state to make decisions.
+//
+// Sampling is one of the last steps when starting a new span, such that the
+// sampler can inspect most of the state of the span to make a decision.
+//
+// Implementations must be safe for concurrent use by multiple goroutines.
+type TracesSampler interface {
+	Sample(ctx SamplingContext) bool
+}
+
+// A SamplingContext is passed to a TracesSampler to determine a sampling
+// decision.
+type SamplingContext struct {
+	Span   *Span // The current span, always non-nil.
+	Parent *Span // The parent span, may be nil.
+}
+
+// The TracesSamplerFunc type is an adapter to allow the use of ordinary
+// functions as a TracesSampler.
+type TracesSamplerFunc func(ctx SamplingContext) bool
+
+var _ TracesSampler = TracesSamplerFunc(nil)
+
+func (f TracesSamplerFunc) Sample(ctx SamplingContext) bool {
+	return f(ctx)
+}
+
+// fixedRateSampler is a TracesSampler that samples root spans randomly at a
+// uniform rate.
+type fixedRateSampler struct {
```

Comment:
We'll probably export this sampler in a later release to allow people to compose logic in their custom sampler.